### ISSUE TITLE
Earth Mover's Distance Loss

### ISSUE THREAD

**Issue by meder411** [pytorch/pytorch#16599]
## 🚀 Feature
(Approximate) Earth Mover's Distance Loss Function

## Motivation
This is a common function used in deep learning papers for point cloud generation and distribution matching.

## Pitch
Can this be added to the loss function library?

## Additional context
There are a few open source implementations out there already. I have updated [this one](https://github.com/fxia22/pointGAN/tree/74b6c432c5eaa1e0a833e755f450df2ee2c5488e/emd) for PyTorch 1.0 according to the CUDA extension tutorial. There are some bugs though that go beyond my expertise (or understanding) when it comes to larger point sets. My version [can be found here](https://github.com/meder411/PyTorch-EMDLoss).

**Comment by mustafa-qamaruddin** [comment]
It would be amazing if this feature gets integrated to the master branch. Being implemented in CUDA and C should make it very efficient. I am training a WGAN, so I would definitely vote for this feature.

**Comment by julioasotodv** [comment]
Indeed, it would be great to have a de-facto implementation of Wasserstein distance in Pytorch for WGANs and other models (such as modified VAEs)

**Comment by izdeby** [comment]
@soumith

**Comment by mustafa-qamaruddin** [comment]
I have just implemented WGAN as per the WGAN source code associated with the published paper. It did not require any sort of custom loss function. Actually, it used Lipschitz inequality and instead of

loss.backward()

It passed the label as the gradient with:

output.backward(-1)  # fake
output.backward(+1) # real

It works like magic. This was very smart and the results were vivid.

So I guess EMD loss has nothing to do with WGAN.

**Comment by gchanan** [comment]
Just getting the loss should be fairly easy, but an efficient algorithm may be more challenging.

**Comment by gchanan** [comment]
we would accept a PR that does the exact computation; talk to us if you want to do a more complex implementation.
